### PR TITLE
forge: resolve and validate the screen/flow/test artifact graph

### PR DESCRIPTION
**Pull request by JMBattista**
## Summary
EPIC #404 → Screens and Flows as UI Feature Kinds → User Story 6 (flow-lint validates the screen/flow/test graph in app CI) → Slice 1: Resolve the Screen/Flow/Test Artifact Graph

Adds a deterministic, smithy-state-free graph resolver (`src/flow-lint.ts`) that discovers durable screen annotations and flow definitions from the repo tree and validates the cross-reference rules — dangling `screens:` references, missing paired test bodies, orphan executable test bodies, and duplicate flat `ScreenId`/`FlowId` values. This is the load-bearing behavior behind the future `smithy flow-lint` command; Slice 1 stops at the resolver + validator so failure and success cases are provable through fixtures before CLI/CI ergonomics (Slices 2–3) are added.

## Spec debt & uncertainty
- SD-010 (open): Fully driver-neutral enumeration of orphan test bodies is reliable only within the declared `test-body:` set or an explicitly supplied flow-test root; a general machine-readable stub marker is deferred. Scoped here — the candidate universe is the declared `test-body:` paths, and orphan detection runs only against a supplied flow-test root or the conventional stub location (`maestro/flows`), reporting `not-run` otherwise rather than blind-scanning the repo.
- SD-005 (inherited): Fidelity of `import`-mode structure derivation — owned by US5; flow-lint validates the confirmed graph after it exists.
- SD-006 (inherited): Whether `SC`/`FL` nodes are always atomic — owned by US3; validation is agnostic to how producing tasks were sliced.
- SD-007 (inherited): Build-phase coverage honesty — flow-lint checks graph integrity, not whether every brief state has executable coverage.
- SD-008 (inherited): Visual-intent honesty under the non-blocking gate — owned by US4; not validated here.
- Assumption: `test-body` is a driver-neutral repo-relative path that may live anywhere in the test tree (per `smithy.helper-flow-definition`); the declared path is treated as the authoritative contract rather than inferring a test layout.
- Verification note: the resolver is not yet wired into the CLI bundle (that is Slice 2), so `npm run build` exercises the existing CLI only; the new module is covered by its own vitest suite. The repo toolchain requires Node ≥ 20.12 (`styleText`); I ran the suite under Node 22 after the initial Node 18 install produced a mismatched native binding.

## Validation
build ✅ · typecheck ✅ · test ✅ (10 passed, `src/flow-lint.test.ts`)

### CODE REVIEW

File: src/flow-lint.ts [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+export function discoverFlowGraph(
+  rootOrSubpath: string,
+  options: DiscoverFlowGraphOptions = {},
+): FlowGraph {
+  const root = resolveArtifactRoot(rootOrSubpath);
+  const screens = discoverScreens(root);
+  const flows = discoverFlows(root);
+  const declaredTestBodyPaths = uniqueSorted(
+    flows.map((flow) => normalizeRelativePath(flow.testBodyPath)).filter(Boolean),
+  );
```
[the remark this answers is not in the document]
Comment (JMBattista) [reply src/flow-lint.ts:85]: Fixed in c533bb8: `validateFlowGraph` now groups flows by declared `test-body` and emits a `duplicate-test-body` finding when two `.flow.md` files share one body, so the 1:1 flow↔body contract is enforced instead of being collapsed during discovery. The dedup in `declaredTestBodyPaths` remains (it feeds the orphan set), but uniqueness is now checked against the un-deduped flow list.

File: src/flow-lint.ts [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+          type: 'missing-screen',
+          flowId: flow.id,
+          flowPath: flow.path,
+          screenId,
+        });
+      }
+    }
+
+    const absoluteTestBodyPath = path.join(graph.root, flow.testBodyPath);
+    if (!isFile(absoluteTestBodyPath)) {
```
[the remark this answers is not in the document]
Comment (JMBattista) [reply src/flow-lint.ts]: Fixed in c533bb8: test-body targets are now resolved via a within-root helper and a path that escapes the artifact root (absolute or `..` traversal) is reported as `missing-test-body`. One correction on the mechanism: `path.join(root, absPath)` does *not* drop the root (that's `path.resolve`) — it produces `root/absPath`, which simply doesn't exist and was already flagged. The real escape vector was `..` traversal, which the containment guard now closes.

File: src/flow-lint.ts [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+}
+
+function discoverOrphanTestBodies(
+  root: string,
+  declaredTestBodyPaths: Set<string>,
+  explicitFlowTestRoot?: string,
+): OrphanScanResult {
+  const scopedRoots = explicitFlowTestRoot
+    ? [resolveScopedRoot(root, explicitFlowTestRoot)]
+    : conventionalFlowTestRoots(root);
```
[the remark this answers is not in the document]
Comment (JMBattista) [reply src/flow-lint.ts]: Fixed in c533bb8: `resolveScopedRoot` now checks `isDirectory` before touching `realpathSync` (so a missing or broken-symlink `flowTestRoot` degrades to a not-run orphan scan instead of throwing) and rejects a resolved root that escapes the artifact root, so `../..` also yields not-run rather than scanning outside. Covered by two new fixtures.

File: src/flow-lint.ts [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+  return Array.from(new Set(values)).sort();
+}
+
+function relativePath(root: string, target: string): string {
+  return normalizeRelativePath(path.relative(root, target));
+}
+
+function normalizeRelativePath(value: string): string {
+  return value.split(path.sep).join('/').replace(/^\.\//, '');
+}
```
[the remark this answers is not in the document]
Comment (JMBattista) [reply src/flow-lint.ts:350]: Fixed in c533bb8: `normalizeRelativePath` now splits on both `/` and `\` so a Windows-authored `test-body` in frontmatter (e.g. `maestro\flows\AddTitle.yaml`) resolves identically on a POSIX CI host.

File: src/flow-lint.ts [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+    ? [resolveScopedRoot(root, explicitFlowTestRoot)]
+    : conventionalFlowTestRoots(root);
+
+  if (scopedRoots.length === 0) {
+    return { status: 'not-run', roots: [], paths: [] };
+  }
+
+  const orphanPaths = scopedRoots.flatMap((scanRoot) =>
+    listFiles(scanRoot).map((filePath) => relativePath(root, filePath)),
+  ).filter((testBodyPath) => !declaredTestBodyPaths.has(testBodyPath));
```
[the remark this answers is not in the document]
Comment (JMBattista) [reply src/flow-lint.ts]: Fixed in c533bb8: orphan scanning now skips any path with a hidden segment before comparing against the declared set, so `.gitkeep`, `.DS_Store`, and files under dot-directories no longer surface as `orphan-test-body` findings. Covered by a new fixture.

File: src/flow-lint.test.ts [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+  it('scopes orphan test-body discovery to the explicit flow-test root', () => {
+    write('design/screens/Library.design.md', screen('Library'));
+    write('design/flows/AddTitle.flow.md', flow('AddTitle', ['Library'], 'ui-flows/AddTitle.yaml'));
+    write('ui-flows/AddTitle.yaml', '# flow\n');
+    write('ui-flows/RemovedFlow.yaml', '# orphan\n');
+    write('tests/e2e/unrelated.spec.ts', 'test("unrelated", () => {});\n');
+
+    const graph = discoverFlowGraph(root, { flowTestRoot: 'ui-flows' });
+
```
[the remark this answers is not in the document]
Comment (JMBattista) [reply src/flow-lint.test.ts:93]: Fixed in c533bb8: added fixtures asserting that a missing `flowTestRoot` and a root-escaping `flowTestRoot` (`../..`) both yield `{ status: 'not-run' }`, locking in deterministic behavior ahead of the Slice 2 CLI wiring.
